packages/monitoring-system/data-processor/test/unit/data-processors/mocks/octokit-middleware.ts

```diff
@@ -0,0 +1,173 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import {Octokit} from '@octokit/rest';
+import {resolve} from 'path';
+import {
+  OctokitRequestParser,
+  GitHubActionType,
+  GitHubObjectType,
+  GitHubActionDetails,
+  OctokitRequestOptions,
+} from './octokit-request-parser';
+
+/**
+ * Parsed details of GitHub action
+ */
+interface GitHubActionDetailsStrict {
+  type: GitHubActionType;
+  value: string;
+  destObjType: GitHubObjectType;
+  destObjId: string | number;
+  repoName: string;
+  repoOwner: string;
+}
+
+export interface MockResponse {
+  type: 'resolve' | 'reject';
+  value: {} | [];
+}
+
+interface MockResponses {
+  [type: string]: {
+    [value: string]: {
+      [dstObjType: string]: {
+        [destObjId: string]: {
+          [repoName: string]: {
+            [repoOwner: string]: MockResponse;
+          };
+        };
+      };
+    };
+  };
+}
+
+/**
+ * A middleware to intercept outgoing Octokit requests and
+ * return predefined responses instead.
+ */
+export class OctokitMiddleware {
+  private static PATH_TO_PLUGIN =
+    './build/test/unit/data-processors/mocks/mock-octokit-plugin.js';
+
+  private static instance: OctokitMiddleware;
+  private mockResponses: MockResponses = {};
+
+  /**
+   * Returns a mock Octokit instance with OctokitMiddleware
+   * injected into it.
+   */
+  public static getMockOctokit(): Octokit {
+    const OctokitWithMiddleware = Octokit.plugin(
```

Comment:
Would you ever want to chain other plugins on as well? e.g. if you wanted the logger plugin added

Reply:
You might want to but it would be hard to support because plugins usually do one of four things:

1. Do something before a request
2. Do something after a request
3. Wrap a request
4. Add new methods

Since this plugin is wrapping the request, adding more hooks before/after might cause undefined behaviour. Also, if a plugin adds new methods, those might not be intercepted by the middleware.